addons/mail/static/src/discuss/voice_message/common/voice_recorder.js

```diff
@@ -0,0 +1,189 @@
+            recording: false,
+            elapsed: "00 : 00",
+        });
+        this.notification = useService("notification");
+        this.userSettings = useService("mail.user_settings");
+        this.voiceMessageService = useState(useService("discuss.voice_message"));
+        this.config = {
+            // 128 or 160 kbit/s – mid-range bitrate quality
+            bitRate: 128,
+        };
+        onWillUnmount(() => {
+            this.cleanUp();
+        });
+    }
+
+    filename() {
+        return (
+            "Voice-" +
+            new Date().toISOString().split("T")[0] +
+            "-" +
+            Math.floor(Math.random() * 100000) +
+            ".mp3"
+        );
+    }
+
+    async startRecording() {
+        if (this.state.isActionPending) {
+            return;
+        }
+        this.state.isActionPending = true;
+        if (!this.microphone) {
+            try {
+                this.microphone = await browser.navigator.mediaDevices.getUserMedia({
+                    audio: this.userSettings.audioConstraints,
+                });
+            } catch {
+                this.notification.add(
+                    sprintf(_t('"%(hostname)s" needs to access your microphone'), {
+                        hostname: window.location.host,
+                    }),
+                    { type: "warning" }
+                );
+                this.state.isActionPending = false;
+                return;
+            }
+        }
+        this.state.elapsed = "00 : 00";
+        this.props.onchangeRecording?.();
+        this.state.recording = true;
+        this.context = new AudioContext();
+
+        await this.context.audioWorklet.addModule("/discuss/voice/worklet_processor");
+        this.processor = new AudioWorkletNode(this.context, "processor");
+        this.processor.port.onmessage = (e) => {
+            if (this.state.recording && !this.startTimeStamp) {
+                this.startTimeStamp = e.timeStamp;
+            }
+            if (!this.startTimeStamp) {
+                return;
+            }
+            const elapsedSeconds = Math.floor((e.timeStamp - this.startTimeStamp) / 1000);
+            const second = elapsedSeconds % 60;
+            const minute = Math.floor(elapsedSeconds / 60);
+            this.state.elapsed =
+                (minute < 10 ? "0" + minute : minute) +
+                " : " +
+                (second < 10 ? "0" + second : second);
+            if (elapsedSeconds > 55 && elapsedSeconds < 60) {
+                this.state.limitWarning = true;
+            }
+            if (elapsedSeconds === 60) {
+                this.notification.add(
+                    _t("The duration of voice messages is limited to 1 minute."),
+                    { type: "warning" }
+                );
+                this.stopRecording();
+            }
+            if (!e.data) {
+                return;
+            }
+            this.encoder.encode(e.data);
```

Comment:
Would it be possible to import the library inside the audio worklet and do the encoding and all the math above in the audio thread, then tranfer the `finalBuffer` (array of int8arrays so should be transferable) from the audio thread to the main JS thread?

Reply:
That's a good idea, I'll try that :)

Reply:
Not doing it in the end, because worklet cannot import any other JS files, and we can't change `lamejs` lib file as per the license.